File: options_trader_node/src/scanner/filters.js
export function toFloat(value, fallback = 0) {
  const n = Number(value);
  return Number.isFinite(n) ? n : fallback;
}

export function toInt(value, fallback = 0) {
  const n = Number.parseInt(value, 10);
  return Number.isFinite(n) ? n : fallback;
}

export function dteFromExpiration(expirationDate, now = new Date()) {
  const exp = new Date(`${expirationDate}T00:00:00`);
  const diffMs = exp.getTime() - now.getTime();
  return Math.max(Math.floor(diffMs / (24 * 60 * 60 * 1000)), 0);
}

export function spreadPct(bid, ask) {
  const mid = (bid + ask) / 2;
  if (mid <= 0) return 1;
  return (ask - bid) / mid;
}

export function normalizeZeroToOne(value, lower, upper) {
  if (upper <= lower) return 0;
  const clipped = Math.max(lower, Math.min(upper, value));
  return (clipped - lower) / (upper - lower);
}

export function passDelta(optionType, delta, min, max) {
  if (optionType === "call") return delta >= min && delta <= max;
  if (optionType === "put") return delta <= -min && delta >= -max;
  return false;
}

File: options_trader/scanner/filters.py
from __future__ import annotations

from datetime import date, datetime


def parse_float(value: object, default: float = 0.0) -> float:
    try:
        return float(value)  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return default


def parse_int(value: object, default: int = 0) -> int:
    try:
        return int(float(value))  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return default


def dte_from_expiration(expiration_date: str, today: date | None = None) -> int:
    base = today or date.today()
    exp = datetime.strptime(expiration_date, "%Y-%m-%d").date()
    return max((exp - base).days, 0)


def spread_pct(bid: float, ask: float) -> float:
    mid = (bid + ask) / 2 if (bid + ask) > 0 else 0.0
    if mid <= 0:
        return 1.0
    return (ask - bid) / mid


def passes_dte(dte: int, dte_min: int, dte_max: int) -> bool:
    return dte_min <= dte <= dte_max


def passes_delta(option_type: str, delta: float, delta_min: float, delta_max: float) -> bool:
    if option_type == "call":
        return delta_min <= delta <= delta_max
    if option_type == "put":
        return -delta_max <= delta <= -delta_min
    return False


def passes_open_interest(open_interest: int, minimum: int) -> bool:
    return open_interest >= minimum


def passes_spread_pct(spread: float, maximum: float) -> bool:
    return spread <= maximum


def normalize_zero_to_one(value: float, lower: float, upper: float) -> float:
    if upper <= lower:
        return 0.0
    clipped = max(lower, min(value, upper))
    return (clipped - lower) / (upper - lower)
